### PR TITLE
graduate SuspendJob to beta

### PR DESCRIPTION
**Pull request by adtac**
#### What type of PR is this?

/kind api-change

#### What this PR does / why we need it:

1. Graduates SuspendJob to beta.
2. Adds a label to two existing Job metrics.
3. Changes the behaviour of syncJob so that each operation only either creates pods or deletes pods (but never both at the same time). This is okay because this will happen only in indexed Jobs with failed indexes, where we'll want to do both. And in such situations, we don't necessarily need to do both pod creations and deletions in the same sync.

KEP update: https://github.com/kubernetes/enhancements/pull/2646

#### Which issue(s) this PR fixes:

xref: https://github.com/kubernetes/enhancements/issues/2232

#### Does this PR introduce a user-facing change?

```release-note
Suspend Job feature graduated to beta
Added the "action" label to Job controller sync metrics job_sync_total and job_sync_duration_seconds
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig apps
/cc @alculquicondor @ahg-g

### CODE REVIEW

File: pkg/apis/batch/types.go
Comment (alculquicondor) [inline pkg/apis/batch/types.go:203]: Requires the SuspendJob feature gate to be enabled (enabled by default).

The "Defaults to false" sentence should be before the beta sentence. Also better put the beta sentence in a separate paragraph.

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:912]: can you put the label values as constants in the metrics package we a comment for what they mean?

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:816]: add to the description that we are changing the sync to only do deletions or creations but not both and why that is acceptable.

File: pkg/controller/job/metrics/metrics.go
Comment (alculquicondor) [inline pkg/controller/job/metrics/metrics.go:43]: add this change to the release note

File: test/e2e/apps/job.go
Comment (alculquicondor) [inline test/e2e/apps/job.go:116]: Did you add it to the infra repo? Don't forget to remove it.

Comment (adtac) [reply test/e2e/apps/job.go:116]: yep, once this PR is merged

File: pkg/apis/batch/types.go
Comment (ahg-g) [inline pkg/apis/batch/types.go:205]:
```suggestion
	// This field is beta-level, gated by SuspendJob feature flag.
```

File: pkg/controller/job/job_controller.go
Comment (ahg-g) [inline pkg/controller/job/job_controller.go:816]: add a comment indicating that we either delete or add pods in a cycle, and deleting pods takes precedence.

File: pkg/controller/job/job_controller.go
Comment (ahg-g) [inline pkg/controller/job/job_controller.go:826]: how do the above two conditions ever be possible if wantActive is strictly larger than active in this branch?

Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:826]: they are not possible, but the code was there 🤷‍♂️

File: pkg/controller/job/metrics/metrics.go
Comment (alculquicondor) [inline pkg/controller/job/metrics/metrics.go:73]: why not const?

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:826]: that said, let's remove it.

File: pkg/controller/job/metrics/metrics.go
Comment (logicalhan) [reply pkg/controller/job/metrics/metrics.go:43]: Definitely, it's a breaking metrics change.

Comment (alculquicondor) [reply pkg/controller/job/metrics/metrics.go:43]: Not really breaking considering that this metric didn't hit a release yet.

File: test/e2e/apps/job.go
Comment (adtac) [reply test/e2e/apps/job.go:116]: ref: https://github.com/kubernetes/test-infra/pull/22188

**Comment by alculquicondor** [comment]
maybe still needs rebase?

**Comment by adtac** [comment]
Not sure what's changed, but I had to run `update-generated-swagger-docs.sh` and `update-openapi-spec.sh` instead of `make generated_files` for the changes to be picked up for some reason o.O

Anyway, rebased and squashed.

File: pkg/controller/job/job_controller.go
Comment (ahg-g) [reply pkg/controller/job/job_controller.go:816]: this wasn't addressed

**Comment by ahg-g** [comment]
/lgtm

Just one final nit.

File: pkg/controller/job/job_controller.go
Comment (adtac) [reply pkg/controller/job/job_controller.go:816]: ah, must have missed it. done.

File: pkg/controller/job/job_controller.go
Comment (soltysh) [inline pkg/controller/job/job_controller.go:822]: Any particular reason to drop this safety net?

File: pkg/controller/job/metrics/metrics.go
Comment (soltysh) [inline pkg/controller/job/metrics/metrics.go:34]: I haven't seen approval from the monitoring team on this change, yet.

Comment (alculquicondor) [reply pkg/controller/job/metrics/metrics.go:34]: @logicalhan PTAL

File: pkg/controller/job/job_controller.go
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:822]: but it's already checked in line 820

Comment (ahg-g) [reply pkg/controller/job/job_controller.go:822]: because of the condition above, `wantActive` is strictly larger than `active`

File: pkg/controller/job/metrics/metrics.go
Comment (ahg-g) [reply pkg/controller/job/metrics/metrics.go:34]: @logicalhan can you please review the metrics changes?

File: pkg/controller/job/job_controller.go
Comment (ahg-g) [reply pkg/controller/job/job_controller.go:822]: ha, I should have refreshed my page, I didn't see your replies :)

**Review by janetkuo — APPROVED** [review]
LGTM from my end; need @logicalhan to look at the metrics change still.

**Review by logicalhan — COMMENTED** [review]
/lgtm
From a metrics perspective. When this feature is goes to GA, you may want to consider promoting this metric to STABLE, since presumably people will be alerting off of it.

**Comment by liggitt** [comment]
/approve

for API doc update

**Comment by alculquicondor** [comment]
all parties already LGTMed

/hold cancel

**Pull request merged**
